### ISSUE TITLE
Make s2s as fast as amun

### ISSUE THREAD

**Issue by emjotde** [marian-nmt/marian#96]
For translation of single sentences in a batch, we are getting closer. The difference used to be 50%, now it is around 10-20%. These are numbers for amun from this repo and s2s from current master in marian-train (soon to be updated here). To put this into perspective, Nematus is probably still more than 4 times slower than either of them. Time given in seconds, ignores warm-up time, measured as time from first to last sentence produced, both are WMT test sets. 

On a GTX 1080, translating 2000 ro-en sentences (50,000 target voc):

| tool | 1 GPU | 4 GPUs |
|------|-----------|-------------|
| amun | 131.2s | 31.9s |
| s2s |  149.7s | 40.3s |

On Titan X Pascal, translating 2168 en-de sentences (85,000 target voc):

| tool | 1 GPU | 2 GPUs |
|------|-----------|-------------|
| amun |  100.7s  | 50.7s |
| s2s | 124.2s  | 64.1s |
| s2s unaligned | 149.5s | - |

`s2s unaligned` is the current s2s version in the `marian` repo before update from `marian-train`. This shows nicely how much we gained from the 256-byte aligned memory allocator from https://github.com/marian-nmt/marian-train/issues/50

I believe it is unlikely that s2s can get much faster in this setting without tricks like scheduling. Amun works more locally on the same memory during inference, s2s always goes from one tensor to another (unless I can do something about this during inference). I will soon start working on batched decoding for s2s and possibly on better scheduling in marian in general. 

I wonder if we should start thinking about retiring amun or keep it around as a pure CPU version.

**Comment by hieuhoang** [comment]
Now that the marian-decoder had batching, I thought I'll try it out for speed v. amun. Both use the same model (85k vocab, 10k sentence test set, beam size 12, running on GTX 1080ti).
   Amun: 145 sec
   Marian: 400 sec
Is this a fair comparison?

**Comment by emjotde** [comment]
Add `--max-batch-sort src`

**Comment by tomekd** [comment]
Can you provide numbers for batch-size=1?

**Comment by emjotde** [comment]
Also, use a larger workspace, like `-w 2500` that way no time will be lost on re-allocation when benchmarking.

**Comment by hieuhoang** [comment]
more numbers. marian-dev/marian-decoder with -w 2500 and --maxi-batch-sort is better. 

There's quite a lot of differences, not sure if the search errors are systematic or just due to ties. Thats another thing to check i suppose
![a](https://user-images.githubusercontent.com/691732/33581915-afbe37de-d94a-11e7-9ba0-a36ab59ca55c.png)

**Comment by hieuhoang** [comment]
about a third of sentences are different

**Comment by emjotde** [comment]
Does this also happen with --mini-batch 1 ?

**Comment by hieuhoang** [comment]
yep

**Comment by emjotde** [comment]
weird, model, test files and commands, please.

**Comment by hieuhoang** [comment]
aye aye. WIll bring it to the office to upload tomorrow. Home broadband is crap

**Comment by emjotde** [comment]
Trained with Nematus or Marian?

**Comment by hieuhoang** [comment]
marian a couple weeks ago

**Comment by emjotde** [comment]
@snukky Don't we have regression tests for translation? If nothing changed there recently I would say that Marian should be the authority how to translate with Marian models. Still, will take a look.

**Comment by hieuhoang** [comment]
That's not a sufficient test, the amin tests goes back further. As rico
says, we have to look at model scores and nbest lists

Hieu Hoang
Sent while bumping into things

**Comment by snukky** [comment]
We have regression tests for all: translations, nbest lists and rescoring.

**Comment by emjotde** [comment]
@hieuhoang So here is a quick test you can do yourself, are the scores for the first translation in the nbest list and in marian-scorer the same?

**Comment by hieuhoang** [comment]
here a quick test you can do yourself - run marian decoder with beam-size=10 twice on the same machine, model & test set. The results are likely to be different each time.

Not sure what regression tests you have for it, definitely can't say the marian-decoder is the authority yet

My model is here
   http://thor.inf.ed.ac.uk/~s0565741/download/vocab85k/

**Comment by emjotde** [comment]
Maybe the "amun" type has a bug. I''l have a look.

**Comment by emjotde** [comment]
We are also adding a regression test to compare against nematus now, so that should prevent things like that in the future.

**Comment by emjotde** [comment]
Update. Bug confirmed but so far only for mini-batch size larger than 64, seems to work fine for 64 and below. Output between amun, marian and different marian runs is the same for small mini-batch. For large mini-batch we get currently between 15-20 different sentences per 1000. So there is some kind of search error.

**Comment by emjotde** [comment]
Instability is fixed. One small _syncthreads() missing.

**Comment by hieuhoang** [comment]
did you push to marian? I'm still getting differences between different runs of marian-decoder. It's not as bad as before, about 53 out of 10,000 sentences

**Comment by emjotde** [comment]
I did. Is this with the same settings? I only tested for the first 1000, maybe I missed that.

**Comment by hieuhoang** [comment]
BEAM=12
BATCH=128
MAXIBATCH=$((BATCH*10))

**Comment by emjotde** [comment]
Tried that, get identical results between multiple runs.

**Comment by hieuhoang** [comment]
oops my bad, was running old version

**Comment by hieuhoang** [comment]
right, speed results with up to date code and tweaks
    beam | amun | marian-decoder
   ----------------------------------------
      12    | 120    | 210
       5     |  37     | 102
       1     | 14      | 28
     Loading/unloading only | 1.15 | 3.51
     Beam 12, batch 1 | 504 | 738

The output is different, especially for larger beams. But this can be looked at another time and thread
   $ diff out.amun.beam1 out.marian.beam1 | wc
    132    2558   14294
  $ diff out.amun.beam5 out.marian.beam5 | wc
  10506  191787 1031420
  $ diff out.amun.beam12 out.marian.beam12 | wc
  12922  251890 1356189

**Comment by emjotde** [comment]
Differences out of how many? Also I guess that line number need to be divided by at least 5?

**Comment by hieuhoang** [comment]
10k sentences

**Comment by emjotde** [comment]
Hm, I had again, no differences between amun and marian for same settings and beam-size (12). Can you take another look?

**Comment by emjotde** [comment]
I get two(!) different sentences for 10K.

**Comment by snukky** [comment]
There are results I've obtained for WMT16 en-de model on the WMT16 test set on a single GTX 1080. Total times without loading models (sec) and word per second (wps):

| setting (sec) | amun | marian | nematus |
|:---|---:| ---:|---:|
| WMT16 en-de, beam=6, batch=1 | 101 | 132 | 729 |
| WMT16 en-de, beam=6, batch=64 |  14 | 35 | - |

| setting (wps) | amun | marian | nematus |
|:---|---:| ---:|---:|
| WMT16 en-de, beam=6, batch=1 | 3,681 | 2,808 | 510 |
| WMT16 en-de, beam=6, batch=64 |  26,300 | 10,930 | - |

I used 5 processes for Nematus. All outputs are almost identical (1 sentence is different between nematus and marian, 4 sentences are different between nematus and amun).

**Comment by emjotde** [comment]
@snukky Would you recommend a different setting that 3x source?

@hieuhoang is it possible to switch off logsoftmax/top-k fusion to see how much that contributes and if it would be worth adding to marian-decoder?

**Comment by hieuhoang** [comment]
in god.cpp around line 90
 useFusedSoftmax_ = false;

**Comment by emjotde** [comment]
@snukky Can you switch fusion off in amun and try again?

**Comment by hieuhoang** [comment]
I've rechecked and retested the comparison with CUDA 8 & 9, beam 1, 5, 12, batch 1, 65, 128, # sentences 1000 & 10k. It's close for beam=1 but nowhere near identical.

The only difference left is machine/drivers. You running on hodor with CUDA 8?

**Comment by emjotde** [comment]
Yes. Near identical for 10K sentences, beam-size 12, mini-batch 1,64,128. Difference about 2 sentences.

**Comment by emjotde** [comment]
It doesn't even matter that Roman is running experiments there in parallel, still identical.

**Comment by hieuhoang** [comment]
ok, my worry is that there's a race condition which is exposed by the different driver/GPU. Would like to see if the output matches the output in my environment

**Comment by emjotde** [comment]
But marian is consistent between marian runs? If yes you would need to compare against nematus.

**Comment by hieuhoang** [comment]
yes, i dont know how to run nematus

**Comment by snukky** [comment]
@hieuhoang you can run the following subset of marian regression tests locally to check if you have the same outputs:

    ./run_mrt.sh tests/models

**Comment by hieuhoang** [comment]
Arggg! Sentence normalization! Now they match perfectly

**Comment by snukky** [comment]
I updated all results in the table above including Amun with disabled softmax/top-k fusion.

**Comment by hieuhoang** [comment]
looks about right. Doesnt add a huge amount, but better than a slap in the face with a wet fish

**Comment by emjotde** [comment]
Batched decoding is a bit at a disadvantage in marian, because I have to do a transpose to get the beams continuous. Will have a to take another look at that. Was so close for mini-batch already :)

**Comment by hieuhoang** [comment]
don't count your chickens before they hatch!

**Comment by emjotde** [comment]
You better invest that energy into making marian faster!  :)

**Comment by hieuhoang** [comment]
@snukky has the nematus comparison finished running yet?

**Comment by snukky** [comment]
@hieuhoang added to the table.

**Comment by hieuhoang** [comment]
cheers. No numbers for batches? Thats the more important 1 imo.

**Comment by emjotde** [comment]
Nematus doesn't have batched translation for beam-size >1

**Comment by hieuhoang** [comment]
The original table had more nematus numbers with batches. You might was well share it for completeness, even if it's beam-size=1

**Comment by emjotde** [comment]
I don't think it had. I am not sure we even know how to run batched translation with Nematus.

**Comment by hieuhoang** [comment]
![chart](https://user-images.githubusercontent.com/691732/33831362-aa5a803e-de6f-11e7-8c81-9240ab6daa1b.png)

Looks like a scalability issue in marian

**Comment by emjotde** [comment]
My guess would be this is due to the computational graph in s2s, every small operation is from memory A to memory B, while amun has in-place operations. It might make sense to thing of in-place operations for inference only.

**Comment by emjotde** [comment]
Tomasz pointed out correctly that Marian always keeps at least as many beams around as batch-size. Amun on the other hand drops finished sentences. This is probably the greatest cause and also explains why there is no such large difference for batch-size=1. I am currently not sure how to solve this, but will take a look. I guess one solution would be to have an inference-only version of attention where different dimensions per batch entry can work as in amun. But this gets messy quickly. 

The advantage of Marian's current solution is that with future reinforcement learning I can easily back-propagate through beam-search.

**Comment by hieuhoang** [comment]
they both sound plausible

**Comment by emjotde** [comment]
Eventually, we will fix this in marian-dev

**Comment by emjotde** [comment]
Ha, quite exactly a year later, this seems to be solved :) Need to do benchmarking.
